### PR TITLE
[PT] Introduce GraphBuilderMode

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**
### Changes

Add GraphBuilderMode to build graph of pytorch model based on FunctionHookMode

### Related tickets

154914

### Tests

tests/torch2/function_hook/graph/test_build_graph_mode.py
tests/torch2/function_hook/graph/test_graph_visualisation.py

### CODE REVIEW

File: nncf/experimental/torch2/function_hook/build_graph_mode.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+    def __init__(self, model: nn.Module, hook_storage: HookStorage):
+        """
+        Initialize the GraphBuilderMode.
+
+        :param model: The PyTorch model to which the hooks will be applied.
+        :param hook_storage: Storage for hooks to be executed.
+        """
+        super().__init__(model=model, hook_storage=hook_storage)
+        self.next_node_id: int = 0
+        self.graph: nx.DiGraph = nx.DiGraph()
```
Comment (alexsu52) [inline nncf/experimental/torch2/function_hook/build_graph_mode.py:56]: Please use `MultiDiGraph`

File: nncf/experimental/torch2/function_hook/graph_utils.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
+# Copyright (c) 2024 Intel Corporation
```
Comment (alexsu52) [inline nncf/experimental/torch2/function_hook/graph_utils.py:1]: What do you think about creating nncf/experimental/torch2/function_hook/graph module?

Comment (AlexanderDokuchaev) [reply nncf/experimental/torch2/function_hook/graph_utils.py]: done

**Review by alexsu52 — APPROVED** [review]
LGTM

File: nncf/experimental/torch2/function_hook/graph/build_graph_mode.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+                dtype=buffer.dtype,
+                output_port_id=0,
+                source_node_id=None,
+                name_in_model=name,
+            )
+
+    def __enter__(self) -> GraphBuilderMode:
+        """
+        Overload __enter__ to get correct return type hint.
+        """
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch2/function_hook/graph/build_graph_mode.py:83]: Return value is missed

File: nncf/experimental/torch2/function_hook/graph/build_graph_mode.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+        super().__enter__()
+        return self
+
+    def register_new_node_id(self) -> int:
+        """
+        Return unique id for new node.
+        """
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch2/function_hook/graph/build_graph_mode.py:90]: Return value is missed

File: nncf/experimental/torch2/function_hook/hook_executor_mode.py [hunk 5]
```diff
@@ -363,13 +366,16 @@ def execute_post_hooks(self, output: Any, op_meta: OpMeta) -> Any:
 
             if isinstance(output, list):
                 for idx, value in enumerate(output):
-                    output[idx] = self.hook_storage.execute_post_function_hooks(op_meta.op_name, idx, value)
+                    output[idx] = self.process_post_function_hooks_for_value(value, op_meta, idx)
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch2/function_hook/hook_executor_mode.py:369]: Question: why this change is only applicable for post hooks and does not apply for pre hooks?

**Review by daniil-lyakhov — COMMENTED** [review]
Do we plan to add a graph test for selected architectures as it done by other backends?

File: nncf/experimental/torch2/function_hook/hook_executor_mode.py (hunk 5 above)
Comment (AlexanderDokuchaev) [reply nncf/experimental/torch2/function_hook/hook_executor_mode.py:380]: For pre hook need to not only register tensor and add edge, also need to process input arguments (tesnormeta) and add it to FunctionMeta

File: nncf/experimental/torch2/function_hook/hook_executor_mode.py [hunk 6]
```diff
@@ -416,7 +422,27 @@ def process_model_outputs(self, outputs: Any) -> Any:
...
+        :param name: The name of the input argument.
+        :param value: The value of the input argument.
+        :returns: The processed value after the hook is executed.
+        """
+        return self.hook_storage.execute_pre_function_hooks(name, 0, value)
+
+    @contextmanager
+    def disable(self) -> Iterator[None]:
+        """
+        Temporarily disables the function tracing and execution hooks within a context.
```
[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/experimental/torch2/function_hook/hook_executor_mode.py:451]: what should i write here?

File: nncf/experimental/torch2/function_hook/graph/build_graph_mode.py [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+    def execute_pre_hooks(
+        self, args: Tuple[Any, ...], kwargs: Dict[str, Any], op_meta: OpMeta
+    ) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
+        """
+        Overloaded function to register a model output to the graph.
+
+        :param args: The arguments to the function.
+        :param kwargs: The keyword arguments to the function.
+        :param op_meta: Metadata for the operation.
+        :returns: The modified arguments and keyword arguments after pre-hooks.
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch2/function_hook/graph/build_graph_mode.py:171]: Please refactor `returns` -> `return` here and elsewhere

File: nncf/experimental/torch2/function_hook/hook_executor_mode.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/experimental/torch2/function_hook/hook_executor_mode.py:451]:
```
:return: A context manager which temporary disables the function tracing and execution hooks within a context.
```

[the remark this answers is not in the document]
Comment (AlexanderDokuchaev) [reply nncf/experimental/torch2/function_hook/hook_executor_mode.py:451]: > ```
> :return: A context manager which temporary disables the function tracing and execution hooks within a context.
> ```

```
a = ctx.disable() # a = None 
```

None is not context manager

**Review by daniil-lyakhov — APPROVED** [review]
